### ISSUE TITLE
How to publish notifications about data availability at a web service (other then file server)

### ISSUE THREAD

**Issue by josusky** [wmo-im/GTStoWIS2#34]
In relation with issue #13 we have discussed a list of file transfer protocols to allow in the URL that is part of the notification/message. And we have briefly touched notifications about data that are distributed through a web service such as OGC WMS, WFS, WFC or EDR. I see two possible approaches:

1. We add a new element into the message schema, something like "serviceType" that would distinguish otherwise identically looking notifications containing a base URL starting with "https://". Current (default) value could be then something like `file` or `fileServer`. While an OGC WFC service that provides the same data would have `serviceType` set to something like `OGC-WFC`.
2. We decide (require data providers) to publish notifications related to services separated by topic. Here I see again two possibilities:
2.1. Services would have a separate sub-tree (or even a root element?). Example: `v03.20210318.WIS.OGC-WFS.ca.upperair.aircraft.airep` or `v03.20210318.WIS.OGC-EDR.gb.bracknell.model-regional`.
2.2. One tree would rule all and different services would just have separate "leafs", e.g. `v03.20210318.WIS.ca.CYXL.upperair.aircraft.airep.ca.OGC-WFS`
I would prefer the first one as I expect the sub-tree for services to be less detailed - as some of the details could be deferred to the OGC protocol level.

**Comment by petersilva** [comment]
for 1.  I'm wondering if we need another element in the message.  perhaps just use the existing baseURL, but adding an ogc-wfc scheme:   ogc-wfc://yourogc-wfcserver.com.  I'm thinking we will do something similar for webdav, S3, and IPFS, where they are transported over https, but not the same as plain https.   not saying we should do this, just that I'm not sure why we would not do it that way. more of a question than a suggestion.


For 2.  in the tree we agreed to today, the date is gone.  it would be v03.WIS.OGS-WFS ... or it could be the common tree v03.WIS...   Don't think we know enough to decide between the two. might want some trials with each... The thing is, there are few enough people looking at the tree we have... I think having two trees with less eyes per tree is probably bad, but otoh, perhaps needs for OGC tree are different... would want to see worked out examples.

**Comment by tomkralidis** [comment]
Nice to see this.  Comments:
- I think here we mean WCS, not 'WFC'?
- it would be valuable to have a controlled vocabulary of some sort to normalize the service types.  Options:
  - OGC Naming Authority/Registry ([example](http://defs.opengis.net/vocprez/object?uri=http://www.opengis.net/def/serviceType/ogc/)).  Having said this, a value like `http://www.opengis.net/def/serviceType/ogc/wms` probably wouldn't work well
  - OSGeo [Link Types tables](https://github.com/OSGeo/Cat-Interop/blob/master/LinkPropertyLookupTable.csv) (defacto/community driven). Informal but used in numerous FOSS4G tools. So an example would be: `v03.WIS.OGC:WFS`

**Comment by petersilva** [comment]
in link types tables...	http://www.ogcnetwork.net/WNS is a broken link ... looked like the most relevant one... was disappointed ;-(   We could modify the existing tree to use a vocabulary that is more appropriate... the current words used in the hierarchy is just trasnscribing/transliterating WMO-386.  Perhaps someone has already done such work when figuring out GRIB or BUFR?

**Comment by petersilva** [comment]
fwiw... could just have a sub-topic for each type   v03.WIS.OGC.WFS  .... in mqtt  v03/WIS/OGC/WFS ...

**Comment by tomkralidis** [comment]
> in link types tables... http://www.ogcnetwork.net/WNS is a broken link ..

It's a URI and does not have to resolve.  In our case here we would use the identifiers as the leaves of the subtopic.

**Comment by petersilva** [comment]
OK... let's take for granted someone approved of something like v02.WIS.OGC.WFS.   what does the topic tree after WFS look like? so... example:
```
http://hpfx.collab.science.gc.ca/~pas037/WMO_Sketch/20210322T23/WIS/ca/canadian_met_centre/surface/land/synop/intermediate/ca/SICN36_CWAO_222100_RRD_ba22f8c4b823b2063b874707bb3766d4.txt	
```
In the GTStoWIS proposal you would have two topics identifing the originating centre of the data, 
ca.canadian_met_centre... then you could add the type of data you are looking for: observation/... but it is confusing here because at the same level is 'surface'... (which can have obs and forecasts in the same tree hmm... )
then the topic tree is identifying the type of layer you want to query... 

ca/canadian_met_centre/surface/land/synop/intermediate/ca/

means a query for an observation reported by the Canadian met centre for a land station, using a synoptic report for an intermediate time, over an area in Canada.   That's as far as the other topic tree goes... then you add something else to be more OGCish?   It does not say what format the report should be in (TAC, BUFR, GRIB, GEOTIFF, CSV?) (we specifically avoided identifying for format so that reports in various formats would end up in the same file folder) It does not say anything about the geographical location.. beyond what country it  is in, and we are lucky about that because some headers encode things like "between 0-90n and 0-90w" so the geo contribution is, shall we say, limited.

Can this partial tree be useful?  I'm wondering if you can come up with a hiearchy that can be compatible with the (modified as necessary) table based tree, and then the OGC stuff goes a few levels deeper, adding specificity.
that way people learn a single tree, and just stop earlier when they are looking at files.

**Comment by antje-s** [comment]
An additional field/attribute in the message like "serviceType" to distinguish between 
- notifications for web services to inform that new data services are available and
- notifications for services that can be used to download new data 
seems useful to me. Then for integrity a value of 0 could be allowed for specific serviceTypes to indicate additionally that it is not a filedownload service

**Comment by antje-s** [comment]
In the case of web services, a large number of different "products" can be available (e.g. through filtering) and the notification is intended more as information for the customer that, for example, the data from the new model run is now available. 
Since the parameters are not set, no data retrieval can take place directly, but the notification could be used as a start signal for the execution of scripts on the customer side.

**Comment by petersilva** [comment]
updating this issue based on activities elsewhere... How I understand (or perhaps misunderstand) this issue:

We don't have any working proposals yet, just vague suggestions.  It would be weird to publish something that says: please do an OGC-API request now, (always sending out the same URL... because we are essentially saying the answer has changed.)

Another aspect here is that when we think about REST type access, then we are saying that query/response should be reproducible, which essentially means that the API request should produce a static response... if that is true, then it should map to a static file name. Having relPath is then a good sanity check, preserving interop and round-tripping between WAF and OGC-API. If we accept that, then using retPath (retrieval path override) covers the entire case, no extension/modification to format needed.

The other aspect is that I'm not sure that this discussion has not morphed into the suggestion for metadata notifications to be published, with metadata having its own topic tree. The MD record would say something like: any time you want "This" issue this OGC-API call.   Has this discussion moved over there?

**Comment by kaiwirt** [comment]
Not being part of the discussion but my understanding of the problem:

When we have a WebService which can be used to query a subset of available model data (for example cut out an area of a global model), then different users will download different data through the same WebService.
Now we need the Pub/Sub to inform the users, when new model data is available (such as a new model run). However this needs to be supported by the format of the Pub/Sub messages as file-related items like file size or integrity value don't apply here.

**Comment by tomkralidis** [comment]
Agree that this would be a different message type, pointing to the relevant collection.

**Comment by petersilva** [comment]
If an area that is a cut-out area of a global model is produced, then it should clearly be available as a file.  Downloading a complete file using WAF is far more efficient than passing through OGC-API, so the WAF should be preferred for those who will consume the entire product.  For those who will consume some overlapping subset using  an OGC-API query... ok I see that need.... interesting...

**Comment by petersilva** [comment]
doesn´t this loop back to having geo-graphical information in the message, either by a field, or (my preference) as part of the file name convention?

**Comment by kaiwirt** [comment]
My use-case is, that we do not produce different cut-out areas of a model available as file. Instead we make available the global modal as such and allow the user to cut-out areas of their choosing using some sort of WebService (for example OGC). 

Thus User A cuts out Area a and User B cuts out Area b and we have Area c

Again: We do not want to produce a and b and c beforehand and fixed. Instead, they are dynamicly generated by the WebService when queried.
However we still need MQP messages to tell A and B and C, that they can now download their products with fresh model data when the global model is available.

The geo-graphical information in that case is not known beforehand and thus can't be part of the message.

**Comment by petersilva** [comment]
With the existing payload format, you don´t need to produce the file, only define the name that the file would have if you did create it storing that in relPath (relative path.) The retPath (retrieval path) overrides it, so subscribers download using the retPath override, and if they download and don´t operate their own OGC-API end point, they have a standard file name to assign to it.  The OGC-API query would be in the retPath, and the WAF compatible filename (that you do not need to write) would be in relPath.

Is that enough?

**Comment by kaiwirt** [comment]
Suppose i have a webservice at URL https:<SomeURL> which takes parameters <parameter>,<level>,<lat>,<lon>

Then a request could be https://<SomeURL>/?parameter=T&level=2&lat=40..60&lon=-10..10

However as the data producer does not know the exact parameters the users will choose the message can only contain <SomeURL> and the possible parameters and values need to be describes somewhere (metadata? openapi?, but thats off topic from the notifications)

**Comment by petersilva** [comment]
yes... the options etc... are out of scope... would have to go somewhere else... but the dimensions of the subset product you have produced would corresponde to a single set of values.... so the query url that corresponds to retrieving the entire subset product would go in retPath. One would ideally construct a corresponding file name to go in relPath.

**Comment by antje-s** [comment]
If we distribute it as a message with a download URL, how can client script differentiate that no automatic download should take place here? Should such messages be distributed under a specific topic and thus be distinguishable from the download messages? 
An automatic download of the whole base product would just cancel out the idea of data reduction before transfer and not only the really needed data would be transferred...

**Comment by josusky** [comment]
I agree with Kai and Antje that the coordinates and other service specific parameters cannot be part of the URL (except when they are static, which would be unusual for a service). In general, the service end-point shall be encoded in the "baseURL" + "retPath" and be treated as opaque (clients/subscribers) should not need to decode it. The "relPath", that is conceptually a product identifier, could in the case of a web service, end with a bugus filename that might convey information about model run - just to differentiate it from a similar message about the previous model run output. This "relPath" could be then similar/identical as would be used to inform about availability of the model output on a file server.

Now coming back to the original question, how to distinguish an ordinary file server from an (OGC) web service?
1. In the "baseURL" element, using a non-standard URI scheme, e.g. "ogc-wfc://myservice.example.com/".
2. In the "baseURL" by making it more complex, i.e. a "link" object instead of a string with a URL.
e.g. 
```
    {
      "rel": "service",
      "type": "application/vnd.ogc_wfs+xml",
      "href": "https://service.example.com/wfs/"
    }
```
3. Somewhere else in the notification JSON document, in an additional (optional) attribute.
4. In the topic, by having a subtree reserved for services.

**Comment by petersilva** [comment]
------------------

going back to the static symbolic URLs... I would expect that any time an OGC server is announcing that some new product, field, layer is available,  It is going to have a geographical and temporal extent.  Specifying that extent is giving the boundaries of the new product, which gives hints for subscribers who can select subsets from the product offerred. The url is conceptually the extent of the entire product being offerred, but clients would always use their judgement to download.  We could provide some hint that downloading the entire extent is not what is expected.

If we do not provide any geo extent in the message, then when someone is publishing, as in one example subsets of a global model, the user has choices:
* download  all subsets and discard theones that aren't relevant.
* Issue an OGC request any time any product is announced, and at some point it will cover the area of interest.

I don't think we can usefully avoid giving geographic and temporal extent in the message.  My preferred location for that information is in the relPath.

----------------

Going back @josusky's point...  I see the problem... all four options sound reasonable at the moment... I don't know how to decide between them yet.

For option 2:   #71 is asking for a payload mimetype...

It can just be "datamimetype" : "application/vnd.ogc_wfs+xml"

And the client will just need to know how to handle that... hmm....

**Comment by josusky** [comment]
I had NWP data in mind, where the data for the whole area (and a specific forecast hour) is available in one go. But yes, I can imagine a service where data appears with some geospatial granularity and to make it more extreme, the covered regions are kind of random. For such a complex case we cannot put it into the topic ... I mean technically we could, but encoding coordinates into a string that is otherwise static seems strange to me. And we want to refer to the topic in the metadata, so it should not contain variable elements. The base URL has the same problem. Therefore, I would prefer to have the geographic bounding box (or some more complex geometry?) as another optional parameter. BTW, it would also need a definition of the projection to be able to unambiguously interpret the coordinates (unless we fix it to, let's say, WGS 84).
